[Addresses page]: Integrate API for address data

Msarawan opened the pull request:
Integrate the design and implementation of the address page into the top-level entry on the Broker details page.

![Screenshot from 2024-05-22 18-46-09](https://github.com/artemiscloud/activemq-artemis-self-provisioning-plugin/assets/99248895/b9fd6737-e924-4417-b907-41d231bfa604)

lavocatt commented on src/addresses/Address.component.tsx:
```diff
@@ -0,0 +1,58 @@
+  TableColumn,
+  VirtualizedTable,
+} from '@openshift-console/dynamic-plugin-sdk';
+import { Address } from './Address.container';
```
The `Address` type will be imported from the generated code.

lavocatt commented on src/addresses/Address.container.tsx:
```diff
@@ -0,0 +1,44 @@
+import { AuthContext, K8sResourceCommon } from '../utils';
+import { useJolokiaServiceGetAddresses } from '../openapi/jolokia/queries';
+
+export type Address = {
+  name: string;
+};
```
You can remove this as we will rely on the code generation to obtain this type.

lavocatt commented on api-server/config/openapi.yml:
```diff
@@ -810,3 +810,12 @@ components:
       items:
         type: object
         additionalProperties: true
+    BrokersAddressResponse:
+      type: array
+      items:
+        type: object
+      required:
+        - name
+      properties:
+        name:
+          type: string
```
Let's change this to:
```
    Address:
      type: object
      required:
        - name
      properties:
        name:
          type: string
```

lavocatt commented on src/addresses/AddressRow.tsx:
```diff
@@ -0,0 +1,27 @@
+  TableData,
+  TableColumn,
+} from '@openshift-console/dynamic-plugin-sdk';
+import { Address } from './Address.container';
```
The Address will be imported from the codegen

lavocatt commented on src/brokers/broker-details/BrokerDetails.container.tsx:
```diff
@@ -20,8 +20,9 @@ import { useTranslation } from '../../i18n';
-  QueuesContainer,
-  TopicsContainer,
+  AddressContainer,
+  // QueuesContainer,
+  // TopicsContainer,
```
It's better to not leave commented lines

Msarawan replied:
yes, i will remove and update the PR.
Thanks!

lavocatt commented on src/brokers/broker-details/BrokerDetails.container.tsx:
```diff
@@ -142,12 +143,18 @@ const BrokerDetailsPage: FC = () => {
+            <AddressContainer />
+          </Tab>
+          {/* <Tab eventKey={3} title={<TabTitleText>{t('queues')}</TabTitleText>}>
             <QueuesContainer />
           </Tab>
           <Tab eventKey={4} title={<TabTitleText>{t('topics')}</TabTitleText>}>
             <TopicsContainer />
-          </Tab>
+          </Tab> */}
```
same here for the commented lines.

lavocatt reviewed (commented):
You are making good progress :rocket: 

During our pair programming session we've introduced an important change in the `openapi.yaml` definition. However the way we wrote it was slightly suboptimal.

See we've declared a `BrokersAddressResponse` type that was an Array of properties and used that as the return value of the `addresses` endpoint. As a result the codegen produced a very generic Typescript type:
`export type BrokersAddressResponse = Array<Record<string, any>>;`

Instead we should have declared an `Address` type and say that the `addresses` endpoint would return an array of `Address` objects. This will make the codegen create a much more useful type for us:

```
export type Address = {
  name: string;
};
```

This looks exactly like the type that you created manually in the `Address.container.tsx` file!

At the moment, we still have full control over the `api-server` and we can make it return any type that is convenient to us, so let's leverage that.

Something to keep in mind is that we shouldn't use useEffects when we do things that are purely within React. I've added a pointer to the documentation, it's worth reading.

Msarawan commented:
> You are making good progress 🚀
> 
> During our pair programming session we've introduced an important change in the `openapi.yaml` definition. However the way we wrote it was slightly suboptimal.
> 
> See we've declared a `BrokersAddressResponse` type that was an Array of properties and used that as the return value of the `addresses` endpoint. As a result the codegen produced a very generic Typescript type: `export type BrokersAddressResponse = Array<Record<string, any>>;`
> 
> Instead we should have declared an `Address` type and say that the `addresses` endpoint would return an array of `Address` objects. This will make the codegen create a much more useful type for us:
> 
> ```
> export type Address = {
>   name: string;
> };
> ```
> 
> This looks exactly like the type that you created manually in the `Address.container.tsx` file!
> 
> At the moment, we still have full control over the `api-server` and we can make it return any type that is convenient to us, so let's leverage that.
> 
> Something to keep in mind is that we shouldn't use useEffects when we do things that are purely within React. I've added a pointer to the documentation, it's worth reading.

sure, Thanks @lavocatt !

Msarawan commented on api-server/api/controllers/api_impl.ts:
```diff
@@ -78,10 +78,17 @@ export const getAddresses = (
+          return { name: name };
+        });
+        res.json(returnValue);
+        console.log('returnValue', returnValue);
```
sure

Msarawan commented on api-server/config/openapi.yml:
```diff
@@ -456,7 +456,7 @@ paths:
           content:
             application/json:
               schema:
-                $ref: '#/components/schemas/BrokersResponse'
+                $ref: '#/components/schemas/BrokersAddressResponse'
```
yeah, make sense to me. thanks!

lavocatt commented on src/addresses/Address.container.tsx:
```diff
@@ -1,33 +1,26 @@
-        };
-      });
+    if (isSuccess && rawAddresses) {
+      const AddressData = getAddressData(rawAddresses);
```
One last thing to fix, now that you've updated the yaml, there's no need to transform `rawAddresses` to something else. Just send it to the `addressData` parameter of the `Addresses` component.

Also it's an opportunity to rename `rawAddresses` to `addrresses` since it's not raw data that needs to be parsed anymore.

Msarawan replied:
okay sure, i will update.

gaohoward commented:
The page looking good to me. I think we may add some links to each address to let user navigate to address details or some action items to make this page more useful. But I think that subjects to further discussions before doing that, and in more PRs.

Msarawan commented:
> The page looking good to me. I think we may add some links to each address to let user navigate to address details or some action items to make this page more useful. But I think that subjects to further discussions before doing that, and in more PRs.

Thanks @gaohoward . Yes, that part i continue with my upcoming PR. where onclick on each address we move to the specific address details page and we can show the users some useful details about each address.
